okhttp/src/main/java/okhttp3/AsyncDns.java

```diff
@@ -0,0 +1,66 @@
+package okhttp3;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import javax.annotation.Nullable;
+
+public final class AsyncDns implements Dns {
+
+  private final ExecutorService executor;
+  private final Dns dns;
+
+  AsyncDns(Builder builder) {
+    this.executor = builder.executor;
+    this.dns = builder.dns;
+  }
+
+  public CompletableFuture<List<InetAddress>> lookupAsync(String hostName) {
```

Comment:
My major problem here is that I don't think this is the correct return type of a OkHttp wide single instance Dns implementation.  Just considering the dual stack world (https://en.wikipedia.org/wiki/Happy_Eyeballs) if we want to do this right, then it problem means a stream of InetAddresses e.g. IPv4 might return more quickly than IPv6.

Dns over Http already does this but blocks until both are ready.